Relative time queries

marvelm opened the pull request:
# Context
I would like to be able to create queries where the current time is injected by Parse.

For example:
```javascript
const recentInactiveQ = new Parse.Query(Parse.Installation);
recentInactiveQ.greaterThan('createdAt', { $relativeTime: '2 weeks ago' });
recentInactiveQ.lessThan('updatedAt', { $relativeTime: '3 days ago' });

Parse.Push.send({
  where: recentInactiveQ,
  data: {
    alert: "Here's what you missed this week...",
    url: "https://example.org/this-weeks-content"
  },
});
```

## The future
This type of query can be serialized and re-executed at any time. I plan on adding support for recurring push notifications and relative time queries are essential for this feature.  
It would look something like this (#4105):
```javascript
Parse.Push.createCampaign({
  where: recentInactiveQ,
  data: {
    alert: "Here's what you missed this week...",
    url: "https://example.org/this-weeks-content"
  },
  interval: 'daily',
  sendTime: '17:30',
});
```

flovilmart commented:
I've added all of you as @marvelm is working with me @ampme on those features.

montymxb commented on spec/ParseQuery.spec.js:
```diff
@@ -3105,7 +3105,59 @@ describe('Parse.Query testing', () => {
+      })
+      .then((results) => {
+        expect(results.length).toBe(0);
+      })
```
Considering that we have a test for getting at least one of each object, and one for neither, can we add one that returns both objects? Something using `q.greaterThan('ttl', { $relativeTime: '3 days ago' });` is what comes to mind.

montymxb commented on src/Adapters/Storage/Mongo/MongoTransform.js:
```diff
@@ -533,6 +533,104 @@ function transformTopLevelAtom(atom, field) {
+  parts = parts.filter((part) => part !== '');
+
+  const future = parts[0] === 'in';
+  const past = parts[parts.length - 1] === 'ago';
```
Hmm, wondering if these should be case insensitive. We could lowercase the entire string in advance before processing.

flovilmart replied:
lowercased on https://github.com/parse-community/parse-server/pull/4289/files#diff-82f0219ca721372e8f2bdbc45c83bbdcR537

montymxb replied:
Cool

montymxb commented on src/Adapters/Storage/Mongo/MongoTransform.js:
```diff
@@ -533,6 +533,104 @@ function transformTopLevelAtom(atom, field) {
+  // strip the 'ago' or 'in'
+  if (future) {
+    parts = parts.slice(1);
+  } else if (past) {
```
Do we need to do a check for `past`? Unless there's something else we're planning to add you could just use `else`.

montymxb commented on src/Adapters/Storage/Mongo/MongoTransform.js:
```diff
@@ -533,6 +533,104 @@ function transformTopLevelAtom(atom, field) {
+    switch(interval) {
+    case 'day':
+    case 'days':
+      seconds += val * 24 * 60 * 60;
```
We can just use 2880 rather than performing the multiplication each time, but you could leave a comment to detail it if needed.

montymxb commented on src/Adapters/Storage/Mongo/MongoTransform.js:
```diff
@@ -533,6 +533,104 @@ function transformTopLevelAtom(atom, field) {
+
+    case 'hour':
+    case 'hours':
+      seconds += val * 60 * 60;
```
Same as above with 3600

montymxb commented on src/Adapters/Storage/Mongo/MongoTransform.js:
```diff
@@ -533,6 +533,104 @@ function transformTopLevelAtom(atom, field) {
+    case 'seconds':
+      seconds += val;
+      break;
+
```
Just a thought, but we could support shorthand interval values like `sec`, `min`, `hr` and their plurals. Not necessarily needed however.

montymxb commented on src/Adapters/Storage/Mongo/MongoTransform.js:
```diff
@@ -533,6 +533,104 @@ function transformTopLevelAtom(atom, field) {
+  let seconds = 0;
+  for (const [num, interval] of pairs) {
+    const val = Number(num);
+    if (!Number.isInteger(val)) {
```
With only whole integer values allowed perhaps we should add a test that expects to fail when someone tries something like `2.5 days`.

flovilmart replied:
From the code, as interpolated to seconds, it seems that it should work with non Integer values no?

montymxb replied:
Makes sense it would work, but we would want to test for it if we decide to allow it (or disallow it as currently).

flovilmart replied:
let's stick with integers for now.

flovilmart commented on src/Adapters/Storage/Mongo/MongoTransform.js:
```diff
@@ -565,9 +663,22 @@ function transformConstraint(constraint, field) {
     case '$exists':
     case '$ne':
-    case '$eq':
+    case '$eq': {
```
we may enter on some weird expectations here: `$ne { $relativeTime: "1 day ago" }` would you expect all object that were created the full day before to be excluded or just the one that matches perfectly the current time?

flovilmart replied:
@montymxb what do you think? Should we limit to $lt/$lte/$gt/$gte to be safe?

montymxb replied:
Hmm, it would make the most sense with the query methods you've mentioned above, but not sure if we should simply rule out $ne/$eq. Although they may not be the most effective query methods with relative date matching they _could_ still work, especially if we allow setting the `now` date in the future (but still not that great).

I would just say exclude those, but if someone expects it to work what would we return? Just an empty response like the query matched nothing, or provide some sort of error clarifying that equal/notEqual ops are not allowed using this param, or another approach perhaps.

flovilmart replied:
Or throw an error, that relativeDates operators don’t work with those operators

montymxb replied:
That would work for me 👍 . Simplifying the initial implementation may be best, we can consider it later if someone actually needs to use it.

marvelm commented:
Thanks for reviewing.

I had to rebase against master because of a failing test. I've addressed the issues raised in the review in the commits after "'Add integration test for multiple results' f8ce556"

flovilmart commented:
@montymxb this all look good to me!

montymxb commented:
Changes look good, but the CI looks out of date. I'm going to bring this up to date with the master to retrigger build/coverage checking. I'm expecting it to come up as a ✓ (or close) but I want to see for certain before I give this an approval.
